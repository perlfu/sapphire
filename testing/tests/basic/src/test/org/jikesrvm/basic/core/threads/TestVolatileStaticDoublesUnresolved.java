/*
 *  This file is part of the Jikes RVM project (http://jikesrvm.org).
 *
 *  This file is licensed to You under the Eclipse Public License (EPL);
 *  You may not use this file except in compliance with the License. You
 *  may obtain a copy of the License at
 *
 *      http://www.opensource.org/licenses/eclipse-1.0.php
 *
 *  See the COPYRIGHT.txt file distributed with this work for information
 *  regarding copyright ownership.
 */
package test.org.jikesrvm.basic.core.threads;

class TestVolatileStaticDoublesUnresolved extends XThread {

  public static void main(String[] args) {
    for (int i = 0; i < 5; i++) {
      TestVolatileStaticDoublesUnresolved tvsdu = new TestVolatileStaticDoublesUnresolved(doubleValues[i], i);
      tvsdu.start();
    }
    XThread.say("bye");
    XThread.outputMessages();
  }

  static double[] doubleValues = { 0.1d, 5934093850936.32940348509376d, 123456.7891011d, -0.00000100023500008d, 1.37470092062392304E17d};

  double d;

  TestVolatileStaticDoublesUnresolved(double d, int threadNumber) {
    super("VSDU" + threadNumber);
    this.d = d;
  }

  void performTask() {
    int errors = 0;
    for (int i = 0; i < 10000000; i++) {
      double td = StaticVolatileDoubleField.vd;
      StaticVolatileDoubleField.vd = d;
      boolean acceptableValue = false;
      for (int j = 0; j < 5; j++) {
        acceptableValue = acceptableValue || (td == doubleValues[j]);
      }
      if (!acceptableValue) errors++;
    }
    tsay(errors + " errors found");
  }

  private static class StaticVolatileDoubleField {
    static volatile double vd = doubleValues[0];
  }

}
