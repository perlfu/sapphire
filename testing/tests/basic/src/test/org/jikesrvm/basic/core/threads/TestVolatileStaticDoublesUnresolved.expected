Thread 0 Main: bye
Thread 1 VSDU0: creating
Thread 1 VSDU0: starting
Thread 1 VSDU0: run starting
Thread 1 VSDU0: 0 errors found
Thread 1 VSDU0: run finishing
Thread 2 VSDU1: creating
Thread 2 VSDU1: starting
Thread 2 VSDU1: run starting
Thread 2 VSDU1: 0 errors found
Thread 2 VSDU1: run finishing
Thread 3 VSDU2: creating
Thread 3 VSDU2: starting
Thread 3 VSDU2: run starting
Thread 3 VSDU2: 0 errors found
Thread 3 VSDU2: run finishing
Thread 4 VSDU3: creating
Thread 4 VSDU3: starting
Thread 4 VSDU3: run starting
Thread 4 VSDU3: 0 errors found
Thread 4 VSDU3: run finishing
Thread 5 VSDU4: creating
Thread 5 VSDU4: starting
Thread 5 VSDU4: run starting
Thread 5 VSDU4: 0 errors found
Thread 5 VSDU4: run finishing
